system/core/pages.php

```diff
@@ -360,7 +360,7 @@ function loadPage(RequestContext $O_O) {
 					if ($return !== null) {
 						echo json_encode($return);
 					} else {
-						echo '<script language="javascript" type="text/javascript">window.top.window.uploadResponse(\''.$response.'\');</script>';
+						echo '[{"error":"'.$response.'"}]';
```

Comment:
This could run into problems with escaping (e.g. when there is a `"` in `$response`). I would suggest:

    if ($return === null)
        $return = [['error' => $response]];

And then let the existing `json_encode` handle (dropping the if before it).

There is some additional refactoring that could maybe be done here, but some of this code might be replaced by Laurens' backend changes anyway.

Reply:
Will do :)